go/libraries/doltcore/env/actions/infer_schema.go

```diff
@@ -0,0 +1,419 @@
+	ColMapper StrMapper
+	// FloatThreshold is the threshold at which a string representing a floating point number should be interpreted as
+	// a float versus an int.  If FloatThreshold is 0.0 then any number with a decimal point will be interpreted as a
+	// float (such as 0.0, 1.0, etc).  If FloatThreshold is 1.0 then any number with a decimal point will be converted
+	// to an int (0.5 will be the int 0, 1.99 will be the int 1, etc.  If the FloatThreshold is 0.001 then numbers with
+	// a fractional component greater than or equal to 0.001 will be treated as a float (1.0 would be an int, 1.0009 would
+	// be an int, 1.001 would be a float, 1.1 would be a float, etc)
+	FloatThreshold float64
+	// KeepTypes is a flag which tells the inferrer, that if a column already exists in the ExistinchSch then use it's type
+	// without modification.
+	KeepTypes bool
+}
+
+// InferSchemaFromTableReader will infer a tables schema.
+func InferSchemaFromTableReader(ctx context.Context, rd table.TableReadCloser, pkCols []string, args *InferenceArgs) (schema.Schema, error) {
+	pkColToIdx := make(map[string]int, len(pkCols))
+	for i, colName := range pkCols {
+		pkColToIdx[colName] = i
+	}
+
+	inferrer := newInferrer(pkColToIdx, rd.GetSchema(), args)
+
+	rdProcFunc := pipeline.ProcFuncForReader(ctx, rd)
+	p := pipeline.NewAsyncPipeline(rdProcFunc, inferrer.sinkRow, nil, inferrer.badRow)
+	p.Start()
+
+	err := p.Wait()
+
+	if err != nil {
+		return nil, err
+	}
+
+	if inferrer.rowFailure != nil {
+		return nil, inferrer.rowFailure
+	}
+
+	return inferrer.inferSchema()
+}
+
+type inferrer struct {
+	sch        schema.Schema
+	pkColToIdx map[string]int
+	impArgs    *InferenceArgs
+
+	colNames  []string
+	colCount  int
+	colType   []map[types.NomsKind]int
+	negatives []bool
+
+	rowFailure *pipeline.TransformRowFailure
+}
+
+func newInferrer(pkColToIdx map[string]int, sch schema.Schema, args *InferenceArgs) *inferrer {
+	colColl := sch.GetAllCols()
+	colNames := make([]string, 0, colColl.Size())
+
+	_ = colColl.Iter(func(tag uint64, col schema.Column) (stop bool, err error) {
+		colNames = append(colNames, col.Name)
+		return false, nil
+	})
+
+	colCount := len(colNames)
+	colType := make([]map[types.NomsKind]int, colCount)
+	negatives := make([]bool, colCount)
+	for i := 0; i < colCount; i++ {
+		colType[i] = make(map[types.NomsKind]int)
+	}
+
+	return &inferrer{sch, pkColToIdx, args, colNames, colCount, colType, negatives, nil}
+}
+
+func (inf *inferrer) inferSchema() (schema.Schema, error) {
+	cols := make([]schema.Column, 0, inf.colCount)
+	pkCols := make([]schema.Column, 0, inf.colCount)
+	existingCols := inf.impArgs.ExistingSch.GetAllCols()
+
+	tag := uint64(0)
+	for i, name := range inf.colNames {
+		if mappedName, ok := inf.impArgs.ColMapper.Map(name); ok {
+			name = mappedName
+		}
+
+		_, partOfPK := inf.pkColToIdx[name]
+		typeToCount := inf.colType[i]
+		hasNegatives := inf.negatives[i]
```

Comment:
I think this is probably the wrong call. Given that we can't really change the type of a column and preserve history, we should be more permissive with unsigned ints, and make them signed by default (maybe controllable with a flag?)

Reply:
we talked about this.  it is signed be default.